### PR TITLE
Persist/load amp state during training

### LINKED ISSUE #3875: Properly checkpoint models trained with Apex
**Is your feature request related to a problem? Please describe.**

#3866 enables mixed precision training with Apex, but pushes the changes required to properly save and load a model trained with `amp` down the line. As discussed, I am simply opening this issue so I don't forget and will (hopefully) make a PR in the next couple of days.

**Describe the solution you'd like**

Follow the instructions in the Apex docs to checkpoint a model trained with `amp`.

**Additional context**

[Apex docs](https://github.com/NVIDIA/apex#checkpointing).

### PR DESCRIPTION
**Pull request by JohnGiorgi**
# Overview

This is a small modification to `GradientDescentTrainer` that persists the `amp.state_dict` when checkpointing the model and loads it when recovering training state from a checkpoint.

WIP.

## TODO

- [ ] Add tests (help?)
- [x] `Model._load()` should probably call `amp.initialize`, otherwise, a model trained with `amp` might be loaded without it (say, during prediction) which will hurt performance (I think? there would be information loss afaik).
- [x] Related to the point above, it is possible that a user will try to load a model that was trained with `amp`, but `amp` is not installed (i.e. `amp = None`). Should there be a warning or even error in this case?

## Closes

Closes #3875.

### CODE REVIEW

**Comment by dirkgr** [comment]
> Related to the point above, it is possible that a user will try to load a model that was trained with `amp`, but `amp` is not installed (i.e. `amp = None`). Should there be a warning or even error in this case?

Is that a problem? It will just ignore the `amp` state and train anyways, right? If it's not a problem, we don't have to do anything about it.

**Comment by JohnGiorgi** [comment]
> > Related to the point above, it is possible that a user will try to load a model that was trained with `amp`, but `amp` is not installed (i.e. `amp = None`). Should there be a warning or even error in this case?
> 
> Is that a problem? It will just ignore the `amp` state and train anyways, right? If it's not a problem, we don't have to do anything about it.

I guess it is only a problem if, when loading a model trained with mixed precision, _not_ calling `amp.initialize()` impacts model performance in some way. I actually don't know the answer to this

**Comment by dirkgr** [comment]
If it impacts model performance, there should be at least a warning, and maybe more. This is the kind of thing that you might never discover and your model silently underperforms. That's important to avoid.

**Comment by JohnGiorgi** [comment]
> If it impacts model performance, there should be at least a warning, and maybe more. This is the kind of thing that you might never discover and your model silently underperforms. That's important to avoid.

Right, that is my worry as well. I just don't know when or if we expect a model trained with amp but loaded without it (or vice verse) to underperform -- maybe someone who knows more about mixed precision training can chime in.

Should `Model._load()` at least log a warning of some sort so this PR can get merged?

**Comment by dirkgr** [comment]
This is going to be a rare case, so let's not hold up the merge with it. Let's just put a warning in, and we can remove it later if we find it's not important.
